src/chunk_insert_state.c

```diff
@@ -204,6 +205,168 @@ create_chunk_result_relation_info(ChunkDispatch *dispatch, Relation rel, Index r
+									   TupleDesc chunk_desc)
+{
+	bool		found_whole_row;
+
+	Assert(returning_clauses != NIL);
+
+	/* map hypertable attnos -> chunk attnos */
+	returning_clauses = (List *) map_variable_attnos_compat((Node *) returning_clauses, varno, 0, map, map_size,
+															rowtype, &found_whole_row);
+
+	return ExecBuildProjectionInfoCompat(returning_clauses, orig->pi_exprContext,
+										 get_projection_info_slot_compat(orig), NULL, chunk_desc);
+}
+
+static ExprState *
+get_adjusted_onconflictupdate_where(ExprState *hyper_where_state, List *where_quals, AttrNumber *map,
+									int map_size, Index varno, Oid rowtype)
+{
+	bool		found_whole_row;
+
+	Assert(where_quals != NIL);
+	/* map hypertable attnos -> chunk attnos for the hypertable */
+	where_quals = (List *) map_variable_attnos_compat((Node *) where_quals, varno, 0, map, map_size,
+													  rowtype, &found_whole_row);
+	/* map hypertable attnos -> chunk attnos for the "excluded" table */
+	where_quals = (List *) map_variable_attnos_compat((Node *) where_quals, INNER_VAR, 0, map, map_size,
+													  rowtype, &found_whole_row);
+#if PG96
+	return ExecInitExpr((Expr *) where_quals, NULL);
+#else
+	return ExecInitQual(where_quals, NULL);
+#endif
+}
+
+static ProjectionInfo *
+get_adjusted_projection_info_onconflicupdate(ProjectionInfo *orig, List *update_tles, AttrNumber *map,
+											 int map_size, Index varno, Oid rowtype,
+											 TupleDesc chunk_desc)
+{
+	bool		found_whole_row;
+	int			i;
+	ListCell   *lc,
+			   *prev,
+			   *next;
+
+	Assert(update_tles != NIL);
+	/* map hypertable attnos -> chunk attnos for the hypertable */
+	update_tles = (List *) map_variable_attnos_compat((Node *) update_tles, varno, 0, map, map_size,
+													  rowtype, &found_whole_row);
+	/* map hypertable attnos -> chunk attnos for the "excluded" table */
+	update_tles = (List *) map_variable_attnos_compat((Node *) update_tles, INNER_VAR, 0, map, map_size,
+													  rowtype, &found_whole_row);
+
+
+	Assert(list_length(update_tles) == map_size);
+
+	/*
+	 * delete any hypertable tle entries for which there are no chunk
+	 * attributes. This can happen for dropped columns that exist on the
+	 * hyperable but not on the chunks.
+	 */
+	lc = list_head(update_tles);
+	prev = NULL;
+	for (i = 0; i < map_size; i++)
+	{
+		AttrNumber	att = map[i];
+
+		next = lnext(lc);
+
+		if (att == InvalidAttrNumber)
+			update_tles = list_delete_cell(update_tles, lc, prev);
+		else
+			prev = lc;
+		lc = next;
+	}
+
+	/*
+	 * Double-check that the TLE is in the same order as the chunk attributes
+	 * and also fix up the TLE resno after the deletes above
+	 */
+	i = 0;
+	foreach(lc, update_tles)
+	{
+		TargetEntry *t = lfirst(lc);
+		Form_pg_attribute attribute;
+
+		Assert(i < chunk_desc->natts);
+		attribute = chunk_desc->attrs[i];
+
+		if (strcmp(NameStr(attribute->attname), t->resname) != 0)
+			elog(ERROR, "TLE order in onConflictUpdate does not match chunk attribute order");
```

Comment:
Could probably improve this error message. Not clear what `onConflictUpdate` is? Seems to refer to a field in a struct or something. Should probably not go in error message.